simulator/mips/t/unit_test.cpp

```diff
@@ -236,10 +171,621 @@ TEST_CASE( "MIPS32_instr_disasm: Process_Disasm_Branches")
+    CHECK( instr.get_v_dst() == 1);
+}
+
+TEST_CASE( "MIPS64_instr: dsrav")
+{
+    CHECK(MIPS32Instr(0x03298817).get_disasm() == "dsrav $s1, $t1, $t9");
+    MIPS64Instr instr( "dsrav");
+
+    instr.set_v_src( 1, 0);
+    instr.set_v_src( 0, 1);
+    instr.execute();
+    CHECK( instr.get_v_dst() == 1);
+
+    instr.set_v_src( 1, 0);
+    instr.set_v_src( 1, 1);
+    instr.execute();
+    CHECK( instr.get_v_dst() == 0);
+
+    instr.set_v_src( 0x100000000000, 0);
+    instr.set_v_src( 4, 1);
+    instr.execute();
+    CHECK( instr.get_v_dst() == 0x010000000000);
+
+    instr.set_v_src( 0xff00000000000000, 0);
+    instr.set_v_src( 4, 1);
+    instr.execute();
+    CHECK( instr.get_v_dst() == 0xfff0000000000000);
+}
+
+TEST_CASE( "MIPS64_instr: dsrlv")
+{
+    CHECK(MIPS32Instr(0x03298816).get_disasm() == "dsrlv $s1, $t1, $t9");
+    MIPS64Instr instr( "dsrlv");
+
+    instr.set_v_src( 1, 0);
+    instr.set_v_src( 0, 1);
+    instr.execute();
+    CHECK( instr.get_v_dst() == 1);
+
+    instr.set_v_src( 1, 0);
+    instr.set_v_src( 1, 1);
+    instr.execute();
+    CHECK( instr.get_v_dst() == 0);
+
+    instr.set_v_src( 0x10000000, 0);
+    instr.set_v_src( 4, 1);
+    instr.execute();
+    CHECK( instr.get_v_dst() == 0x01000000);
+
+    instr.set_v_src( 0x8000000000000000, 0);
+    instr.set_v_src( 4, 1);
+    instr.execute();
+    CHECK( instr.get_v_dst() == 0x0800000000000000);
+}
+
+TEST_CASE( "MIPS32_instr: jalr")
+{
+    CHECK(MIPS32Instr(0x01208809).get_disasm() == "jalr $s1, $t1");
+    MIPS32Instr instr( "jalr");
+
+    instr.set_new_PC( instr.get_PC() + 4);
+    instr.set_v_src( instr.get_PC() + 8, 0);
+    instr.execute();
+    CHECK( instr.get_new_PC() == instr.get_PC() + 8);
+    CHECK( instr.get_v_dst() == instr.get_PC() + 4);
+
+    instr.set_new_PC( instr.get_PC() + 4);
+    instr.set_v_src( instr.get_PC() + 2, 0);
+    instr.execute();
+    CHECK( instr.get_new_PC() == instr.get_PC() + 4);
+}
+
+TEST_CASE( "MIPS32_instr: jr")
+{
+    CHECK(MIPS32Instr(0x02200008).get_disasm() == "jr $s1");
+    MIPS32Instr instr( "jr");
+
+    instr.set_new_PC( instr.get_PC() + 4);
+    instr.set_v_src( 1024, 0);
+    instr.execute();
+    CHECK( instr.get_new_PC() == 1024);
+
+    instr.set_new_PC( instr.get_PC() + 4);
+    instr.set_v_src( instr.get_PC() + 2, 0);
+    instr.execute();
+    CHECK( instr.get_new_PC() == instr.get_PC() + 4);
+}
+
+TEST_CASE( "MIPS32_instr: mfhi")
+{
+    CHECK(MIPS32Instr(0x00008810).get_disasm() == "mfhi $s1");
+    MIPS32Instr instr( "mfhi");
+    instr.set_v_dst( 0);
+    instr.set_v_src( 0xfee1dead, 0);
+    instr.execute();
+    CHECK( instr.get_v_dst() == 0xfee1dead);
+}
+
+TEST_CASE( "MIPS32_instr: mflo")
+{
+    CHECK(MIPS32Instr(0x00008812).get_disasm() == "mflo $s1");
+    MIPS32Instr instr( "mflo");
+    instr.set_v_dst( 0);
+    instr.set_v_src( 0xfee1dead, 0);
+    instr.execute();
+    CHECK( instr.get_v_dst() == 0xfee1dead);
+}
+
+/*      Not implemented on the register file level
```

Comment:
By the way, I found how to test them without RF.
We implement these instruction via masked register update. If condition is not taken, then "get_mask" return zero and nothing is written to the RF. So, you may check "get_mask" output.

Reply:
Now tests do not pass on the get_mask(). Mask becomes 0 after each operation (all_one at constr.).

Reply:
Currently added test implementation that uses a mask.